### ISSUE TITLE
Unable to correct protocol extension after erroneous attempt in Swift REPL

### ISSUE THREAD

**Issue by Stiivi** [apple/swift#62785]
**Description**

Using REPL, when a wrong conformance to a protocol is specified in an extension, subsequent extensions with a correct conformance are rejected.

**Steps to reproduce**

1. Start `swift repl`
2. Define a protocol: `protocol Boo { func boo() }`
3. Define wrongly a protocol extension conformance: `extension Int: Boo {}`. Correct error is raised:

```
expression failed to parse:
error: repl.swift:2:1: error: type 'Int' does not conform to protocol 'Boo'
extension Int: Boo {}
^

repl.swift:1:21: note: protocol requires function 'boo()' with type '() -> ()'
protocol Boo { func boo() }
                    ^
```

4. Provide correct conformance: `extension Int: Boo { public func boo() {} } `

Results in an error:

```
  2> extension Int: Boo { public func boo() {} } 
expression failed to parse:
error: repl.swift:2:16: error: redundant conformance of 'Int' to protocol 'Boo'
extension Int: Boo { public func boo() {} }
               ^

repl.swift:2:1: note: 'Int' declares conformance to protocol 'Boo' here
extension Int: Boo {}
^
```

**Expected behavior**

No error should be raised after the last correct protocol extension implementation.

Full REPL session:

```
Welcome to Apple Swift version 5.7.2 (swiftlang-5.7.2.135.5 clang-1400.0.29.51).
Type :help for assistance.
  1> protocol Boo { func boo() } 
  2> extension Int: Boo {} 
expression failed to parse:
error: repl.swift:2:1: error: type 'Int' does not conform to protocol 'Boo'
extension Int: Boo {}
^

repl.swift:1:21: note: protocol requires function 'boo()' with type '() -> ()'
protocol Boo { func boo() }
                    ^


  2> extension Int: Boo { public func boo() {} } 
expression failed to parse:
error: repl.swift:2:16: error: redundant conformance of 'Int' to protocol 'Boo'
extension Int: Boo { public func boo() {} }
               ^

repl.swift:2:1: note: 'Int' declares conformance to protocol 'Boo' here
extension Int: Boo {}
^
```

Correct session:

```
Welcome to Apple Swift version 5.7.2 (swiftlang-5.7.2.135.5 clang-1400.0.29.51).
Type :help for assistance.
  1> protocol Boo { func boo() } 
  2> extension Int: Boo { public func boo() {} } 
  3>  
```

**Environment**
- Swift compiler version info: swift-driver version: 1.62.15 Apple Swift version 5.7.2 (swiftlang-5.7.2.135.5 clang-1400.0.29.51) Target: x86_64-apple-macosx12.0
- Xcode version info Xcode 14.2 Build version 14C18
- Deployment target: MacOS 12.6.1 (21G217)

**Comment by AnthonyLatsis** [comment]
This is expected behavior. You cannot *restate* a conformance, regardless of whether the type actually conforms or not. The first conformance failed semantically, but the compiler remembers that you *stated* it syntactically as `: Boo` and won’t allow you to state it again. In a source file, the compiler has immediate knowledge of the `func boo()` implementation in the second extension, so the first conformance statement will succeed; the second one will still produce an error.

**Comment by Stiivi** [comment]
I agree that from one perspective it is an expected behaviour in a source file. We are talking about a different context: REPL - "Repeat-Evaluate-Print-Loop" where it is not a desired behaviour. REPL is an interactive environment (the "Loop" part) and interactive environment should be, at least to some extent, resistant to user errors. This type of error makes the whole REPL session unusable and non-recoverable.

From user's perspective this error is neither unlikely neither has impression of a fatal corruption of the REPL environment.

Let's have the following analogous example:

```swift
let i = 10
let i = "Hello"
```

The above code when executed as a swift source file results in a correct error:

```
wrong.swift:2:5: error: invalid redeclaration of 'i'
let i = "Hello"
    ^
wrong.swift:1:5: note: 'i' previously declared here
let i = 10
    ^
```

However the REPL session succeeds as expected:

```
stefan@Perun:~% swift repl
Welcome to Apple Swift version 5.7.2 (swiftlang-5.7.2.135.5 clang-1400.0.29.51).
Type :help for assistance.
  1> let i = 10
i: Int = 10
  2> let i = "Hello"
i: String = "Hello"
  3>  
```

My point here is, that there are classes of errors should not corrupt the REPL environment and let the user continue. If the REPL environment is corrupted, the REPL should finish and not let the user continue. If the error of this issue corrupts the environment, then the correct solution would be to give up and exit, instead of letting the user to continue.

**Comment by AnthonyLatsis** [comment]
@adrian-prantl How do we not diagnose redeclarations in the REPL? How does that work, I mean.

**Comment by adrian-prantl** [comment]
@augusto2112 can correct me if this is wrong, but I think every line is a new Swift module that imports all the previous ones. You can see what is really happening under the hood by enabling the expression log: `: log enable lldb expr [-f /tmp/expr.log ]`.
